[WIP] Fix type conversion warnings on Windows

jkeiser opened the pull request:
Paying down the Windows warning debt.

Also moved the flags into simdjson-flags.cmake, something I've wanted to do. The general recommendation for cmake projects is to have the top level cmake mostly include other stuff.

jkeiser commented:
Yep, this causes about a 5% stage 1 performance degradation and 1% overall. I'll back off the datatype changes until it recovers.

jkeiser commented:
Once I fixed the warnings emitted by clang on the other end, the performance righted itself.

lemire commented on src/fallback/stage1_find_marks.h:
```diff
@@ -127,7 +127,7 @@ really_inline error_code scan() {
   *next_structural_index = len;
   next_structural_index++;
-  doc_parser.n_structural_indexes = next_structural_index - doc_parser.structural_indexes.get();
+  doc_parser.n_structural_indexes = static_cast<uint32_t>(next_structural_index - doc_parser.structural_indexes.get());
```
Or you can do `static_cast<decltype<doc_parser.n_structural_indexes>>` which has the benefit that you avoid locking yourself into uint32_t. :-)

There may be a more clever way.

jkeiser replied:
Yeah, I was thinking about making a typedef for the size to handle this stuff. But I figured we could add it in when we actually toggle uint32_t to size_t; if I just did it where there were warning fixes we'd likely miss some.

lemire replied:
There is something weird to the C++ culture, where doing "a = b" works all the time even when a and b are different types, and we are totally fine with it... but then, if we are dealing with primitives, we don't trust the magic anymore.

lemire commented:
> Once I fixed the warnings emitted by clang on the other end, the performance righted itself.

Can you elaborate?

jkeiser commented:
> Can you elaborate?

When I changed types, Windows stopped warning, but g++7 (sorry, thought it was clang) picked up some new warnings, particularly signed vs. unsigned comparisons. I changed a couple more types at that point. It may or may not be those changes that fixed perf, but perf does appear to be fixed on my end at least.

lemire commented:
> When I changed types, Windows stopped warning, but g++7 (sorry, thought it was clang) picked up some new warnings, particularly signed vs. unsigned comparisons. I changed a couple more types at that point.

So Visual Studio is fine with == when comparing signed and unsigned... but will complain with inequalities (C4018). Clang should complain in both cases.

I can't imagine how it could massively impact performance... but there you have it.

jkeiser commented:
> I can't imagine how it could massively impact performance... but there you have it.

I'm chalking it up to volatility; but if you wanted to check the perf on another machine, I wouldn't mind the confirmation :)

lemire commented:
The number of instructions in stage 1 is up by a bit less than 1%.

```
$ for i in jsonexamples/*.json; do echo $i ; ./parsemaster $i |grep Instructions && ./parse $i | grep Instructions; done
jsonexamples/apache_builds.json
|    Instructions : 320.0302 per block    ( 99.99%) -   5.0013 per byte    -  51.4833 per structural    -    3.420 per cycle
|    Instructions : 147.4811 per block    ( 46.08%) -   2.3048 per byte    -  23.7253 per structural    -    3.749 per cycle
|    Instructions : 171.8547 per block    ( 53.70%) -   2.6857 per byte    -  27.6463 per structural    -    3.196 per cycle
|    Instructions : 320.5334 per block    ( 99.99%) -   5.0092 per byte    -  51.5643 per structural    -    3.337 per cycle
|    Instructions : 147.9844 per block    ( 46.17%) -   2.3126 per byte    -  23.8063 per structural    -    3.597 per cycle
|    Instructions : 171.8547 per block    ( 53.61%) -   2.6857 per byte    -  27.6463 per structural    -    3.155 per cycle
jsonexamples/canada.json
|    Instructions : 882.1593 per block    (100.00%) -  13.7839 per byte    -  92.7951 per structural    -    3.356 per cycle
|    Instructions : 165.6479 per block    ( 18.78%) -   2.5883 per byte    -  17.4247 per structural    -    3.589 per cycle
|    Instructions : 716.4668 per block    ( 81.22%) -  11.1949 per byte    -  75.3658 per structural    -    3.311 per cycle
|    Instructions : 882.6593 per block    (100.00%) -  13.7917 per byte    -  92.8477 per structural    -    3.435 per cycle
|    Instructions : 166.1479 per block    ( 18.82%) -   2.5961 per byte    -  17.4773 per structural    -    3.455 per cycle
|    Instructions : 716.4668 per block    ( 81.17%) -  11.1949 per byte    -  75.3658 per structural    -    3.433 per cycle
jsonexamples/citm_catalog.json
|    Instructions : 298.9538 per block    (100.00%) -   4.6712 per byte    -  59.3291 per structural    -    3.553 per cycle
|    Instructions : 138.7626 per block    ( 46.42%) -   2.1682 per byte    -  27.5382 per structural    -    3.803 per cycle
|    Instructions : 160.1330 per block    ( 53.56%) -   2.5021 per byte    -  31.7793 per structural    -    3.364 per cycle
|    Instructions : 299.4538 per block    (100.00%) -   4.6790 per byte    -  59.4283 per structural    -    3.356 per cycle
|    Instructions : 139.2626 per block    ( 46.51%) -   2.1760 per byte    -  27.6375 per structural    -    3.630 per cycle
|    Instructions : 160.1330 per block    ( 53.47%) -   2.5021 per byte    -  31.7793 per structural    -    3.153 per cycle
jsonexamples/github_events.json
|    Instructions : 278.9519 per block    ( 99.99%) -   4.3600 per byte    -  60.9908 per structural    -    3.056 per cycle
|    Instructions : 137.3517 per block    ( 49.23%) -   2.1468 per byte    -  30.0309 per structural    -    3.692 per cycle
|    Instructions : 140.2485 per block    ( 50.27%) -   2.1921 per byte    -  30.6643 per structural    -    2.631 per cycle
|    Instructions : 279.5383 per block    ( 99.99%) -   4.3691 per byte    -  61.1190 per structural    -    2.979 per cycle
|    Instructions : 137.9381 per block    ( 49.34%) -   2.1559 per byte    -  30.1591 per structural    -    3.481 per cycle
|    Instructions : 140.2485 per block    ( 50.16%) -   2.1921 per byte    -  30.6643 per structural    -    2.628 per cycle
jsonexamples/gsoc-2018.json
|    Instructions : 178.8457 per block    (100.00%) -   2.7945 per byte    - 122.6200 per structural    -    2.642 per cycle
|    Instructions : 107.3524 per block    ( 60.02%) -   1.6774 per byte    -  73.6028 per structural    -    3.376 per cycle
|    Instructions :  71.4631 per block    ( 39.96%) -   1.1166 per byte    -  48.9964 per structural    -    2.000 per cycle
|    Instructions : 179.6999 per block    (100.00%) -   2.8078 per byte    - 123.2056 per structural    -    2.667 per cycle
|    Instructions : 108.2066 per block    ( 60.21%) -   1.6907 per byte    -  74.1885 per structural    -    3.270 per cycle
|    Instructions :  71.4631 per block    ( 39.77%) -   1.1166 per byte    -  48.9964 per structural    -    2.090 per cycle
jsonexamples/instruments.json
|    Instructions : 359.7496 per block    (100.00%) -   5.6212 per byte    -  45.5827 per structural    -    3.407 per cycle
|    Instructions : 151.2695 per block    ( 42.05%) -   2.3637 per byte    -  19.1669 per structural    -    3.783 per cycle
|    Instructions : 208.0238 per block    ( 57.82%) -   3.2505 per byte    -  26.3580 per structural    -    3.187 per cycle
|    Instructions : 360.2501 per block    (100.00%) -   5.6291 per byte    -  45.6461 per structural    -    3.244 per cycle
|    Instructions : 151.7700 per block    ( 42.13%) -   2.3715 per byte    -  19.2303 per structural    -    3.590 per cycle
|    Instructions : 208.0238 per block    ( 57.74%) -   3.2505 per byte    -  26.3580 per structural    -    3.039 per cycle
jsonexamples/marine_ik.json
|    Instructions : 800.1316 per block    (100.00%) -  12.5021 per byte    -  58.0078 per structural    -    2.756 per cycle
|    Instructions : 189.0420 per block    ( 23.63%) -   2.9538 per byte    -  13.7051 per structural    -    3.510 per cycle
|    Instructions : 611.0559 per block    ( 76.37%) -   9.5478 per byte    -  44.3003 per structural    -    2.587 per cycle
|    Instructions : 800.6316 per block    (100.00%) -  12.5100 per byte    -  58.0441 per structural    -    2.961 per cycle
|    Instructions : 189.5420 per block    ( 23.67%) -   2.9616 per byte    -  13.7414 per structural    -    3.411 per cycle
|    Instructions : 611.0559 per block    ( 76.32%) -   9.5478 per byte    -  44.3003 per structural    -    2.846 per cycle
jsonexamples/mesh.json
|    Instructions : 850.9542 per block    (100.00%) -  13.2971 per byte    -  62.7748 per structural    -    2.773 per cycle
|    Instructions : 182.3537 per block    ( 21.43%) -   2.8495 per byte    -  13.4522 per structural    -    3.754 per cycle
|    Instructions : 668.4616 per block    ( 78.55%) -  10.4454 per byte    -  49.3123 per structural    -    2.591 per cycle
|    Instructions : 851.4543 per block    (100.00%) -  13.3049 per byte    -  62.8117 per structural    -    3.093 per cycle
|    Instructions : 182.8538 per block    ( 21.48%) -   2.8573 per byte    -  13.4891 per structural    -    3.612 per cycle
|    Instructions : 668.4616 per block    ( 78.51%) -  10.4454 per byte    -  49.3123 per structural    -    2.978 per cycle
jsonexamples/mesh.pretty.json
|    Instructions : 519.5725 per block    (100.00%) -   8.1186 per byte    -  83.5491 per structural    -    2.909 per cycle
|    Instructions : 145.7306 per block    ( 28.05%) -   2.2771 per byte    -  23.4340 per structural    -    3.809 per cycle
|    Instructions : 373.7781 per block    ( 71.94%) -   5.8405 per byte    -  60.1048 per structural    -    2.665 per cycle
|    Instructions : 520.0726 per block    (100.00%) -   8.1264 per byte    -  83.6295 per structural    -    3.133 per cycle
|    Instructions : 146.2307 per block    ( 28.12%) -   2.2849 per byte    -  23.5144 per structural    -    3.638 per cycle
|    Instructions : 373.7781 per block    ( 71.87%) -   5.8405 per byte    -  60.1048 per structural    -    2.973 per cycle
jsonexamples/numbers.json
|    Instructions : 712.2451 per block    (100.00%) -  11.1303 per byte    -  83.5338 per structural    -    2.917 per cycle
|    Instructions : 152.4254 per block    ( 21.40%) -   2.3820 per byte    -  17.8768 per structural    -    3.734 per cycle
|    Instructions : 559.1500 per block    ( 78.50%) -   8.7379 per byte    -  65.5785 per structural    -    2.757 per cycle
|    Instructions : 712.7455 per block    (100.00%) -  11.1381 per byte    -  83.5925 per structural    -    3.013 per cycle
|    Instructions : 152.9258 per block    ( 21.46%) -   2.3898 per byte    -  17.9355 per structural    -    3.577 per cycle
|    Instructions : 559.1500 per block    ( 78.45%) -   8.7379 per byte    -  65.5785 per structural    -    2.891 per cycle
jsonexamples/random.json
|    Instructions : 509.5784 per block    (100.00%) -   7.9630 per byte    -  46.1832 per structural    -    3.376 per cycle
|    Instructions : 219.9494 per block    ( 43.16%) -   3.4371 per byte    -  19.9341 per structural    -    3.651 per cycle
|    Instructions : 289.4321 per block    ( 56.80%) -   4.5228 per byte    -  26.2313 per structural    -    3.201 per cycle
|    Instructions : 510.0786 per block    (100.00%) -   7.9708 per byte    -  46.2285 per structural    -    3.332 per cycle
|    Instructions : 220.4495 per block    ( 43.22%) -   3.4449 per byte    -  19.9794 per structural    -    3.530 per cycle
|    Instructions : 289.4321 per block    ( 56.74%) -   4.5228 per byte    -  26.2313 per structural    -    3.199 per cycle
jsonexamples/repeat.json
|    Instructions : 369.2640 per block    ( 99.94%) -   5.7880 per byte    -  64.6303 per structural    -    2.704 per cycle
|    Instructions :   8.5955 per block    (  2.33%) -   0.1347 per byte    -   1.5044 per structural    -    0.497 per cycle
|    Instructions : 201.3034 per block    ( 54.48%) -   3.1553 per byte    -  35.2330 per structural    -    3.199 per cycle
|    Instructions : 159.3652 per block    ( 43.13%) -   2.4980 per byte    -  27.8928 per structural    -    2.867 per cycle
|    Instructions : 369.7697 per block    ( 99.94%) -   5.7960 per byte    -  64.7188 per structural    -    2.762 per cycle
|    Instructions :   8.5955 per block    (  2.32%) -   0.1347 per byte    -   1.5044 per structural    -    0.499 per cycle
|    Instructions : 201.8090 per block    ( 54.55%) -   3.1633 per byte    -  35.3215 per structural    -    3.195 per cycle
|    Instructions : 159.3652 per block    ( 43.07%) -   2.4980 per byte    -  27.8928 per structural    -    2.984 per cycle
jsonexamples/tree-pretty.json
|    Instructions : 353.8667 per block    ( 99.98%) -   5.5298 per byte    -  48.2667 per structural    -    2.871 per cycle
|    Instructions :   3.1574 per block    (  0.89%) -   0.0493 per byte    -   0.4307 per structural    -    0.397 per cycle
|    Instructions : 149.3444 per block    ( 42.20%) -   2.3338 per byte    -  20.3703 per structural    -    3.085 per cycle
|    Instructions : 201.3648 per block    ( 56.89%) -   3.1467 per byte    -  27.4658 per structural    -    3.018 per cycle
|    Instructions : 354.3685 per block    ( 99.98%) -   5.5376 per byte    -  48.3352 per structural    -    2.790 per cycle
|    Instructions :   3.1574 per block    (  0.89%) -   0.0493 per byte    -   0.4307 per structural    -    0.391 per cycle
|    Instructions : 149.8463 per block    ( 42.28%) -   2.3416 per byte    -  20.4387 per structural    -    3.009 per cycle
|    Instructions : 201.3648 per block    ( 56.81%) -   3.1467 per byte    -  27.4658 per structural    -    2.922 per cycle
jsonexamples/twitter_api_compact_response.json
|    Instructions : 394.2532 per block    ( 99.94%) -   6.1828 per byte    -  50.7264 per structural    -    2.482 per cycle
|    Instructions :   9.6582 per block    (  2.45%) -   0.1515 per byte    -   1.2427 per structural    -    0.492 per cycle
|    Instructions : 163.1835 per block    ( 41.37%) -   2.5591 per byte    -  20.9959 per structural    -    3.133 per cycle
|    Instructions : 221.4114 per block    ( 56.13%) -   3.4723 per byte    -  28.4878 per structural    -    2.573 per cycle
|    Instructions : 394.7848 per block    ( 99.94%) -   6.1912 per byte    -  50.7948 per structural    -    2.495 per cycle
|    Instructions :   9.6582 per block    (  2.45%) -   0.1515 per byte    -   1.2427 per structural    -    0.513 per cycle
|    Instructions : 163.7152 per block    ( 41.45%) -   2.5674 per byte    -  21.0643 per structural    -    3.074 per cycle
|    Instructions : 221.4114 per block    ( 56.05%) -   3.4723 per byte    -  28.4878 per structural    -    2.594 per cycle
jsonexamples/twitter_api_response.json
|    Instructions : 321.2845 per block    ( 99.95%) -   5.0342 per byte    -  53.3614 per structural    -    2.585 per cycle
|    Instructions :   6.4059 per block    (  1.99%) -   0.1004 per byte    -   1.0639 per structural    -    0.412 per cycle
|    Instructions : 149.5021 per block    ( 46.51%) -   2.3426 per byte    -  24.8304 per structural    -    3.091 per cycle
|    Instructions : 165.3766 per block    ( 51.45%) -   2.5913 per byte    -  27.4670 per structural    -    2.748 per cycle
|    Instructions : 321.8201 per block    ( 99.95%) -   5.0426 per byte    -  53.4503 per structural    -    2.512 per cycle
|    Instructions :   6.4059 per block    (  1.99%) -   0.1004 per byte    -   1.0639 per structural    -    0.405 per cycle
|    Instructions : 150.0377 per block    ( 46.60%) -   2.3509 per byte    -  24.9194 per structural    -    2.941 per cycle
|    Instructions : 165.3766 per block    ( 51.36%) -   2.5913 per byte    -  27.4670 per structural    -    2.719 per cycle
jsonexamples/twitterescaped.json
|    Instructions : 562.9675 per block    (100.00%) -   8.7967 per byte    -  89.5239 per structural    -    2.758 per cycle
|    Instructions : 139.7079 per block    ( 24.82%) -   2.1830 per byte    -  22.2165 per structural    -    3.394 per cycle
|    Instructions : 423.0808 per block    ( 75.15%) -   6.6109 per byte    -  67.2789 per structural    -    2.601 per cycle
|    Instructions : 563.6222 per block    (100.00%) -   8.8070 per byte    -  89.6280 per structural    -    2.684 per cycle
|    Instructions : 140.3627 per block    ( 24.90%) -   2.1933 per byte    -  22.3207 per structural    -    3.256 per cycle
|    Instructions : 423.0808 per block    ( 75.06%) -   6.6109 per byte    -  67.2789 per structural    -    2.540 per cycle
jsonexamples/twitter.json
|    Instructions : 309.4237 per block    (100.00%) -   4.8350 per byte    -  55.2520 per structural    -    3.193 per cycle
|    Instructions : 156.9502 per block    ( 50.72%) -   2.4525 per byte    -  28.0257 per structural    -    3.586 per cycle
|    Instructions : 152.3142 per block    ( 49.22%) -   2.3800 per byte    -  27.1979 per structural    -    2.884 per cycle
|    Instructions : 309.9900 per block    (100.00%) -   4.8439 per byte    -  55.3531 per structural    -    3.067 per cycle
|    Instructions : 157.5165 per block    ( 50.81%) -   2.4613 per byte    -  28.1268 per structural    -    3.419 per cycle
|    Instructions : 152.3142 per block    ( 49.13%) -   2.3800 per byte    -  27.1979 per structural    -    2.783 per cycle
```

jkeiser commented:
All right, I'll see what can be done to make the conversion changes lighter. I know I've seen performance effects even bigger than this from conversions before!

lemire commented on src/haswell/bitmanipulation.h:
```diff
@@ -36,14 +36,16 @@ really_inline int leading_zeroes(uint64_t input_num) {
   return __popcnt64(input_num);// Visual Studio wants two underscores
+}
 #else
+really_inline int count_ones(uint64_t input_num) {
```
In my tests, this change here (return int) accounts for 50% of the regression...
```
setting it back to uint64_t: 2.4573 per byte
this PR: 2.4614 per byte
master: 2.4525 per byte
```

It is also unclear why you made this change at all given that you did not make this change on westmere.

Now... why would that make *any* difference to the performance. Who knows?

jkeiser replied:
I've had trouble with leading_zeroes and similar functions if the function returns anything other than the actual result type of the intrinsic. Might be I changed a function I didn't need to for some reason, I'll look at it--thanks.

pps83 replied:
maybe try to replace _BitScanReverse64 with __lzcnt64?

lemire commented:
It looks like there is significant performance regression on westmere...

This PR:
```
|- Stage 1
|    Speed        :  21.3537 ns per block ( 55.80%) -   0.3337 ns per byte -   3.8130 ns per structural -    2.997 GB/s
|    Cycles       :  78.8400 per block    ( 56.12%) -   1.2319 per byte    -  14.0780 per structural    -    3.692 GHz est. frequency
|    Instructions : 293.2226 per block    ( 60.35%) -   4.5819 per byte    -  52.3591 per structural    -    3.719 per cycle
|    Misses       :    1411 branch misses ( 46.17%) - 0 cache misses (  0.00%) - 28063.00 cache references
```


Master:
```
|- Stage 1
|    Speed        :  20.4545 ns per block ( 53.62%) -   0.3196 ns per byte -   3.6524 ns per structural -    3.129 GB/s
|    Cycles       :  75.5329 per block    ( 53.94%) -   1.1803 per byte    -  13.4875 per structural    -    3.693 GHz est. frequency
|    Instructions : 287.3537 per block    ( 59.86%) -   4.4902 per byte    -  51.3111 per structural    -    3.804 per cycle
|    Misses       :    1180 branch misses ( 38.22%) - 15 cache misses ( 34.95%) - 30217.00 cache references
```

I am using twitter.json.

lemire commented:
Now for haswell...

This PR:

```
|- Stage 1
|    Speed        :  12.3367 ns per block ( 44.72%) -   0.1928 ns per byte -   2.2029 ns per structural -    5.187 GB/s
|    Cycles       :  45.5554 per block    ( 45.10%) -   0.7118 per byte    -   8.1346 per structural    -    3.693 GHz est. frequency
|    Instructions : 157.5196 per block    ( 50.81%) -   2.4614 per byte    -  28.1274 per structural    -    3.458 per cycle
|    Misses       :    1046 branch misses ( 34.08%) - 0 cache misses (  0.00%) - 29428.00 cache references
```

Master:

```
|- Stage 1
|    Speed        :  11.8489 ns per block ( 44.58%) -   0.1851 ns per byte -   2.1158 ns per structural -    5.401 GB/s
|    Cycles       :  43.7570 per block    ( 44.96%) -   0.6837 per byte    -   7.8134 per structural    -    3.693 GHz est. frequency
|    Instructions : 156.9533 per block    ( 50.72%) -   2.4525 per byte    -  28.0263 per structural    -    3.587 per cycle
|    Misses       :     853 branch misses ( 32.25%) - 0 cache misses (  0.00%) - 29371.00 cache references
```

lemire commented:
For ARM.

This PR:
```
|- Stage 1
|    Speed        :  94.9702 ns per block ( 38.79%) -   1.4840 ns per byte -  16.9583 ns per structural -    0.674 GB/s
|    Cycles       : 285.0251 per block    ( 56.62%) -   4.4538 per byte    -  50.8953 per structural    -    3.001 GHz est. frequency
|    Instructions : 207.3422 per block    ( 50.43%) -   3.2399 per byte    -  37.0239 per structural    -    0.727 per cycle
|    Misses       :    3176 branch misses ( 19.10%) - 9944 cache misses ( 39.93%) - 184879.00 cache references
```

Master:
```
|- Stage 1
|    Speed        :  89.1355 ns per block ( 37.57%) -   1.3928 ns per byte -  15.9164 ns per structural -    0.718 GB/s
|    Cycles       : 285.3369 per block    ( 56.48%) -   4.4587 per byte    -  50.9510 per structural    -    3.201 GHz est. frequency
|    Instructions : 204.3433 per block    ( 50.07%) -   3.1931 per byte    -  36.4884 per structural    -    0.716 per cycle
|    Misses       :    3370 branch misses ( 20.06%) - 9935 cache misses ( 39.91%) - 185197.00 cache references
```

lemire commented:
So something is slowing us down in stage 1 but stage 2 is entirely unaffected performance-wise.

jkeiser commented:
I'm abandoning this in favor of #759, which adds conversion warnings on Linux as well, and doesn't blithely change so many underlying types :)
